src/expr/core/src/expr/wrapper/non_strict.rs

```diff
@@ -129,75 +103,14 @@ where
     async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
         Ok(match self.inner.eval(input).await {
             Ok(array) => array,
-            Err(_e) => self.eval_chunk_infallible_by_row(input).await,
-        })
-    }
-
-    async fn eval_v2(&self, input: &DataChunk) -> Result<ValueImpl> {
-        Ok(match self.inner.eval_v2(input).await {
-            Ok(value) => value,
-            Err(_e) => self.eval_chunk_infallible_by_row(input).await.into(),
-        })
-    }
-
-    async fn eval_row(&self, input: &OwnedRow) -> Result<Datum> {
-        Ok(self.eval_row_infallible(input).await)
-    }
-
-    fn eval_const(&self) -> Result<Datum> {
-        self.inner.eval_const() // do not handle error
-    }
-
-    fn input_ref_index(&self) -> Option<usize> {
-        self.inner.input_ref_index()
-    }
-}
-
-/// Similar to [`NonStrict`] wrapper, but does not fallback to row-based evaluation when an error occurs.
-pub(crate) struct NonStrictNoFallback<E, R> {
-    inner: E,
-    report: R,
-}
-
-impl<E, R> std::fmt::Debug for NonStrictNoFallback<E, R>
-where
-    E: std::fmt::Debug,
-{
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("NonStrictNoFallback")
-            .field("inner", &self.inner)
-            .field("report", &std::any::type_name::<R>())
-            .finish()
-    }
-}
-
-impl<E, R> NonStrictNoFallback<E, R>
-where
-    E: Expression,
-    R: EvalErrorReport,
-{
-    pub fn new(inner: E, report: R) -> Self {
-        Self { inner, report }
-    }
-}
-
-// TODO: avoid the overhead of extra boxing.
-#[async_trait]
-impl<E, R> Expression for NonStrictNoFallback<E, R>
-where
-    E: Expression,
-    R: EvalErrorReport,
-{
-    fn return_type(&self) -> DataType {
-        self.inner.return_type()
-    }
-
-    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
-        Ok(match self.inner.eval(input).await {
-            Ok(array) => array,
-            Err(error) => {
-                self.report.report(error);
-                // no fallback and return NULL for each row
+            Err(ExprError::Multiple(array, errors)) => {
+                for error in errors {
+                    self.report.report(error);
+                }
+                array
+            }
+            Err(e) => {
+                self.report.report(e);
```

Comment:
Do we ensure that (almost) all `Expression`s return `Multiple`? Otherwise, all rows will be padded with `NULL` once there's an error.

Reply:
Yes I have checked that. The remaining expressions not generated by `#[function]` are not evaluated by row, except for UDF, which I will fix later.

Reply:
It looks so fragile to me. 🥵